### PR TITLE
Solidity upgrade deployment

### PR DESCRIPTION
**Pull request by tomlinton**
This is ready for early eyes. The general strategy here was to redeploy everything except OUSDProxy and VaultProxy. I compressed two undeployed deploy scripts (`020_aave_v2.js` and `021_buyback_uni_v3.js`).

Writing out long form what this does:

**020_new_governor.js**
- Deploys a new governor contract
- Proposal:
  - Transfer governance of VaultProxy to new governor
  - Transfer governance of OUSDProxy to new governor
  - Transfer governance of the Buyback contract to new governor

**021_solidity_upgrade.js**
- Deploys new OUSD implementation
- Deploys a new VaultAdmin and VaultCore implementation
- Deploys a new Oracle
- Deploys a new Buyback contract
- Transfer governance of new Buyback to the new Governor
- Deploys a new CompoundStrategyProxy
- Deploys a new CompoundStrategy implementation
- Initializes CompoundStrategyProxy with the CompoundStrategy implementation
- Initializes the CompoundStrategy
- Transfers governance of CompoundStrategy from the deployer to the new Governor
- Deploys a new AaveStrategyProxy
- Deploys a new AaveStrategy implementation
- Initializes AaveStrategyProxy with the AaveStrategy implementation
- Initializes the AaveStrategy
- Transfers governance of AaveStrategy from the deployer to the new Governor
- Deploys a new Flipper
- Transfers governance of Flipper to the new Governor
- Proposal:
  - Claim governance of OUSDProxy
  - Claim governance of VaultProxy
  - Claim governance of Buyback
  - Claim governance of old Buyback
  - Claim governance of Flipper
  - Claim governance of CompoundStrategy
  - Claim governance of AaveStrategy
  - Upgrade OUSD implementation
  - Upgrade VaultCore implementation
  - Upgrade VaultAdmin implementation
  - Set trustee address to the newly deployed Buyback contract
  - Transfer OGN from old Buyback contract to governor
  - Transfer OGN from governor to new Buyback contract
  - Set price provider address to the newly deployed Oracle
  - Add CRV as a swap token
  - Add COMP as a swap token
  - Add AAVE as a swap tokenb
  - Approve new Compound strategy on Vault
  - Set Compound strategy as default for USDT and USDC
  - Approve new AAVE strategy on Vault
  - Add AAVE as default DAI strategy on vault
  - Call allocate

**Contract changes**

- I used OpenZeppelin's newer version of Initializable.sol as part of the upgrade which had a different storage layout, so I had to add the `uint256[50] private ______gap;` from the old version back in.
- I had to remove MIN_DRIFT and MAX_DRIFT checks for some tokens in the OracleRouter, as we now have feeds for non stablecoins.
- Changed MAX_ACTIONS on governance to a higher value to be able to do all this in one go.

**Todo**

- Verify the new buyback contract ends up with the right amount of OGN

**Fork debug output**

You can run this on fork by commenting out `  main.skip = () => !(isMainnet || isRinkeby) || isSmokeTest || isFork;` in `utils/deploy.js`.

```
contracts ❯ npx hardhat debug --network localhost

Contract addresses
====================
OUSD proxy:              0x2A8e1E676Ec238d8A992307B495b45B3fEAa5e86
OUSD impl:               0x307a6343A4ecd5dF8F113fb7f1A78D792F81f91C
OUSD:                    0x307a6343A4ecd5dF8F113fb7f1A78D792F81f91C
Vault proxy:             0xE75D77B1865Ae93c7eaa3040B038D7aA7BC02F70
Vault impl:              0xa9A605405144619e8229645FF4Ce1F53A74A69Bb
Vault:                   0x6bd6CC9605Ae43B424cB06363255b061A84DfFD3
VaultCore:               0xa9A605405144619e8229645FF4Ce1F53A74A69Bb
VaultAdmin:              0xb826CBD8aa01BD3Aa43594A21d5Aa7AD70802f41
OracleRouter:            0x7AAbFcF9B64dA442935F1B1aB45f644D63464483
AaveStrategy proxy:      0x3768278a5407A08960D4F817098AA93ec0a3C570
AaveStrategy impl:       0x3525ec52f699CeeA4ae3382e4B4CC0ba6E176f98
AaveStrategy:            0x3525ec52f699CeeA4ae3382e4B4CC0ba6E176f98
CompoundStrategy proxy:  0xE5643a899C0CA15B5aAb5a01D39741B425F5Bcc0
CompoundStrategy impl:   0x2836f32e009F031e3Fa15E8CD1e7277C560d6Fe3
CompoundStrategy:        0x2836f32e009F031e3Fa15E8CD1e7277C560d6Fe3
ThreePoolStrategy proxy: 0x3c5fe0a3922777343CBD67D3732FCdc9f2Fa6f2F
ThreePoolStrategy impl:  0x874c74E6ec318AD0a7e6f23301678a4751d00482
ThreePoolStrategy:       0x874c74E6ec318AD0a7e6f23301678a4751d00482
Governor:                0x20E0c5F61124D184101a0A8d9afaeA69F5dAB907
OGNStaking proxy:        0x501804B374EF06fa9C427476147ac09F1551B9A0
OGNStaking proxy impl:   0x0Ed143d1D6d98f10db7151b66D76aA3956072b5a
OGNStaking:              0x0Ed143d1D6d98f10db7151b66D76aA3956072b5a

Governor
====================
Admin:            0xbe2AB3d3d8F6a32b96414ebbd865dBD276d3d899
PendingAdmin:     0x0000000000000000000000000000000000000000
Delay (seconds):  60
ProposalCount:    1

Governor addresses
====================
OUSD:               0x20E0c5F61124D184101a0A8d9afaeA69F5dAB907
Vault:              0x20E0c5F61124D184101a0A8d9afaeA69F5dAB907
AaveStrategy:       0x20E0c5F61124D184101a0A8d9afaeA69F5dAB907
CompoundStrategy:   0x20E0c5F61124D184101a0A8d9afaeA69F5dAB907
ThreePoolStrategy:  0x830622BDd79CC677eE6594E20bBda5B26568b781

OUSD
=======
name:                    Origin Dollar
symbol:                  OUSD
decimals:                18
totalSupply:             27140871.779627785581217787
vaultAddress:            0xE75D77B1865Ae93c7eaa3040B038D7aA7BC02F70
nonRebasingSupply:       14008362.658658395156223517
rebasingSupply:          13132509.12096939042499427
rebasingCreditsPerToken: 865827551903448092
rebasingCredits:         11370488222558630366148027

Oracle
========
DAI price :  1.0002 USD
USDC price:  1.00026215 USD
USDT price:  1.00036675 USD

Vault Settings
================
rebasePaused:                    false
capitalPaused:                   false
redeemFeeBps:                    50 (0.5%)
trusteeFeeBps:                   0 (0%)
vaultBuffer:                     0.005 (0.5%)
autoAllocateThreshold (USD):     25000.0
rebaseThreshold (USD):           1000.0
maxSupplyDiff:                   5.0%
Price provider address:          0x7AAbFcF9B64dA442935F1B1aB45f644D63464483
Uniswap address:                 0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D
Strategy count:                  4
Asset count:                     3
Strategist address:              0xF14BBdf064E3F67f51cd9BD646aE3716aD938FDC
Trustee address:                 0xA45Fd81a1145a71001861DC85aDa1E43a22C5D56

Vault balances
================
totalValue (USD):        $27160080.78
  DAI:                   9619092.76
  USDC:                  6077919.12
  USDT:                  11463068.90

Vault buffer balances
================
DAI:                     48095.463817224874
USDC:                    30389.595597
USDT:                    57315.344491

Strategies balances
=====================
Aave DAI:        balance=9570997.299627750184647194
Compound USDC:   balance=6047529.523769
Compound USDT:   balance=11405753.553705
ThreePool DAI:   balance=0.0
ThreePool USDC:  balance=0.0
ThreePool USDT:  balance=0.0

Default strategies
============================
DAI     0x3768278a5407A08960D4F817098AA93ec0a3C570
USDC    0xE5643a899C0CA15B5aAb5a01D39741B425F5Bcc0
USDT    0xE5643a899C0CA15B5aAb5a01D39741B425F5Bcc0

Aave strategy settings
============================
vaultAddress:                    0xE75D77B1865Ae93c7eaa3040B038D7aA7BC02F70
platformAddress:                 0xB53C1a33016B2DC2fF3653530bfF1848a515c8c5
rewardTokenAddress:              0x7Fc66500c84A76Ad7e9c93437bFc5Ac33E2DDaE9
rewardLiquidationThreshold:      0
supportsAsset(DAI):              true
supportsAsset(USDC):             false
supportsAsset(USDT):             false

Compound strategy settings
============================
vaultAddress:                    0xE75D77B1865Ae93c7eaa3040B038D7aA7BC02F70
platformAddress:                 0x0000000000000000000000000000000000000001
rewardTokenAddress:              0xc00e94Cb662C3520282E6f5717214004A7f26888
rewardLiquidationThreshold:      0
supportsAsset(DAI):              false
supportsAsset(USDC):             true
supportsAsset(USDT):             true

3pool strategy settings
==============================
vaultAddress:                    0xE75D77B1865Ae93c7eaa3040B038D7aA7BC02F70
platformAddress:                 0xbEbc44782C7dB0a1A60Cb6fe97d0b483032FF1C7
rewardTokenAddress:              0xD533a949740bb3306d119CC777fa900bA034cd52
rewardLiquidationThreshold:      0
supportsAsset(DAI):              true
supportsAsset(USDC):             true
supportsAsset(USDT):             true
```

### CODE REVIEW

**Review by franckc — COMMENTED** [review]
Wow. Nice work! A couple comments/questions but aside from that looks very solid...

File: contracts/deploy/021_solidity_upgrade.js [hunk 1]
```diff
@@ -0,0 +1,322 @@
+const hre = require("hardhat");
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/deploy/021_solidity_upgrade.js:1]: There isn't any reason I can think of. The only important bit is that the new governor contract gets deployed then the next migration uses its address.

File: contracts/deploy/021_solidity_upgrade.js [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    );
+    const cGovernor = await ethers.getContract("Governor");
+
+    // Oracle contract
+    const dOracleRouter = await deployWithConfirmation("OracleRouter");
+    log("Deployed OracleRouter...");
+
+    // Buyback contract
+    const cOldBuyback = await ethers.getContract("Buyback");
+    const cOGN = await ethers.getContractAt("MockOGN", assetAddresses.OGN);
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/deploy/021_solidity_upgrade.js:39]: It's only using it for the transfer function so I think it is OK.

[the remark this answers is not in the document]
Comment (franckc) [reply contracts/deploy/021_solidity_upgrade.js:39]: ah.. yes. nevermind then...

File: contracts/deploy/021_solidity_upgrade.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (franckc) [reply contracts/deploy/021_solidity_upgrade.js:1]: Sound good. thanks for confirming...

**Review by franckc — APPROVED** [review]
LGTM aside from my comment on gas premium/type 2 transaction.

File: contracts/deploy/020_new_governor.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  { deployName: "020_new_governor" },
+  async ({ ethers, deployWithConfirmation }) => {
+    const { guardianAddr } = await hre.getNamedAccounts();
+
+    const cOUSDProxy = await ethers.getContract("OUSDProxy");
+    const cVaultProxy = await ethers.getContract("VaultProxy");
+    const cBuyback = await ethers.getContract("Buyback");
+
+    const oldGovernor = await ethers.getContract("Governor");
+    const dGovernor = await deployWithConfirmation("Governor", [
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/deploy/020_new_governor.js:15]: Ugh can we just remove the gasPrice premium and hope ethersjs figures out reasonable values? 😁

[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/deploy/020_new_governor.js:15]: Looking through the hardhat-deploy code it's not clear to me whats going to happen with an explicitly passed in gasPrice like we have. From some comments in the code ([e.g](https://github.com/wighawag/hardhat-deploy/blob/9d4c7ce458def49ce36eb4a1ad00f15b96168f7f/src/helpers.ts#L87)., although this is for the CLI) it looks like it'll use a legacy transaction. I'd vote for just trying it and seeing what happens.

[the remark this answers is not in the document]
Comment (franckc) [reply contracts/deploy/020_new_governor.js:15]: Looking at the hardhat-deploy `_deploy` method [source code](https://github.com/wighawag/hardhat-deploy/blob/9d4c7ce458def49ce36eb4a1ad00f15b96168f7f/src/helpers.ts#L523) I think you are right that passing gasPrice as we currently do may just still work with hardhat-deploy.

File: contracts/contracts/utils/Initializable.sol [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        _;
+
+        if (isTopLevelCall) {
+            initializing = false;
+        }
+    }
+
+    uint256[50] private ______gap;
```
Comment (DanielVF) [inline contracts/contracts/utils/Initializable.sol:37]: Is this gap new? Is this going to throw off every storage slot after in the vault and in OUSD?

File: contracts/deploy/021_solidity_upgrade.js [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    log("Initialized AaveStrategy...");
+
+    await withConfirmation(
+      cAaveStrategy.connect(sDeployer).transferGovernance(cGovernor.address)
+    );
+    log("Transferred governance of AaveStrategy...");
+
+    // Flipper
+    await deployWithConfirmation(
+      "Flipper",
```
Comment (DanielVF) [inline contracts/deploy/021_solidity_upgrade.js:162]: Did the flipper change?

File: contracts/deploy/021_solidity_upgrade.js [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+          args: [assetAddresses.USDC, cCompoundStrategyProxy.address],
+        },
+        {
+          // Approve AAVE strategy in Vault
+          contract: cVault,
+          signature: "approveStrategy(address)",
+          args: [cAaveStrategyProxy.address],
+        },
+        {
+          // Add AAVE as default DAI strategy
```
Comment (DanielVF) [inline contracts/deploy/021_solidity_upgrade.js:309]: Let's keep this compound for now and use the strategist to prove out the AAVE strat for a bit with a smaller amount before putting lots of millions in first.

File: contracts/deploy/021_solidity_upgrade.js [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+          args: [assetAddresses.COMP],
+        },
+        {
+          // Add AAVE as a swap token
+          contract: cVault,
+          signature: "addSwapToken(address)",
+          args: [assetAddresses.AAVE],
+        },
+        {
+          // Remove old Compound strategy
```
Comment (DanielVF) [inline contracts/deploy/021_solidity_upgrade.js:279]: Do we need a final harvest before running this?

File: contracts/contracts/utils/Initializable.sol (hunk 4 above)
Comment (tomlinton) [reply contracts/contracts/utils/Initializable.sol:37]: See the PR comment for an explanation

File: contracts/deploy/021_solidity_upgrade.js (hunk 5 above)
Comment (tomlinton) [reply contracts/deploy/021_solidity_upgrade.js:162]: Yes it did with the Solidity upgrade and use of `immutable`. We don't have to redeploy it though.

File: contracts/deploy/021_solidity_upgrade.js (hunk 6 above)
Comment (tomlinton) [reply contracts/deploy/021_solidity_upgrade.js:309]: Ok, I'll remove it as a default strategy but leave it as an approved strategy

File: contracts/deploy/021_solidity_upgrade.js (hunk 7 above)
Comment (DanielVF) [reply contracts/deploy/021_solidity_upgrade.js:279]: Likely to be $10K+

File: contracts/deploy/021_solidity_upgrade.js (hunk 5 above)
Comment (DanielVF) [reply contracts/deploy/021_solidity_upgrade.js:162]: Yeah, I think we can just keep the live version in use.

File: contracts/contracts/utils/Initializable.sol (hunk 4 above)
Comment (DanielVF) [reply contracts/contracts/utils/Initializable.sol:37]: Great.

File: contracts/deploy/021_solidity_upgrade.js (hunk 6 above)
Comment (DanielVF) [reply contracts/deploy/021_solidity_upgrade.js:309]: 👍

File: contracts/deploy/021_solidity_upgrade.js (hunk 5 above)
Comment (tomlinton) [reply contracts/deploy/021_solidity_upgrade.js]: Removed ✅

**Pull request merged**